heartbeat/monitors/monitor.go

```diff
@@ -94,13 +94,31 @@ func (e ErrDuplicateMonitorID) Error() string {
 	return fmt.Sprintf("monitor ID %s is configured for multiple monitors! IDs must be unique values.", e.ID)
 }
 
+// newMonitor Creates a new monitor, without leaking resources in the event of an error.
 func newMonitor(
 	config *common.Config,
 	registrar *pluginsReg,
 	pipelineConnector beat.PipelineConnector,
 	scheduler *scheduler.Scheduler,
 	allowWatches bool,
 	factoryMetadata *common.MapStrPointer,
+) (*Monitor, error) {
+	m, err := newMonitorUnsafe(config, registrar, pipelineConnector, scheduler, allowWatches, factoryMetadata)
+	if m != nil && err != nil {
+		m.Stop()
+	}
+	return m, err
+}
+
+// newMonitorUnsafe is the unsafe way of creating a new monitor because it may return a monitor instance along with an
+// error without freeing monitor resources. m.Stop() must always be called on a non-nil monitor to free resources.
+func newMonitorUnsafe(
+	config *common.Config,
+	registrar *pluginsReg,
+	pipelineConnector beat.PipelineConnector,
+	scheduler *scheduler.Scheduler,
+	allowWatches bool,
+	factoryMetadata *common.MapStrPointer,
```

Comment:
1. Carefull with naming something `Unsafe`. This somewhat implies usage of `unsafe` package, unsafe pointer manipulation and/or potential undefined behavior if someone messed up.

2. The constructor should be self-contained. Never leak resources when an error is to be returned. I know `newMonitor` is still the actual constructor, but we should not encourage devs to use this function just cause it exists. Also devs need to remember to always return m in case of error. As this is somewhat unintuitive, it's easy for us to introduce new leaks here in the future.

All in all I don't mind the presence of `newMonitorUnsafe` as is. For completion an alternative pattern we have in a few places:

```
ok := false
defer cleanup.IfNot(&ok, m.Stop)

// init work

ok = true
return m, nil
```

Maybe it's time to move the [cleanup](https://github.com/elastic/go-txfile/tree/master/internal/cleanup) package to libbeat...

Reply:
Since this is a backport can we defer this to a new master PR?